### PR TITLE
SDIT-4102 Test edge cases for get all offender transfers

### PR DESCRIPTION
**Pull request by mikehalmamoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderTransferScheduleOut.kt [hunk 1]
```diff
@@ -41,9 +39,8 @@ class OffenderTransferScheduleOut(
   )
   var cancellationReasonCode: TransferCancellationReason? = null,
 
-  @OneToOne(mappedBy = "schedule", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
+  @OneToOne(mappedBy = "schedule", cascade = [CascadeType.ALL], fetch = FetchType.EAGER)
   @JoinColumn(name = "EVENT_ID", insertable = false, updatable = false)
-  @NotFound(action = NotFoundAction.IGNORE)
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderTransferScheduleOut.kt:46]: Didn't need the NotFound, it did nothing. The fetch type might now be EAGER but it was still an extra query for Hibernate, so now fetch in the query.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderTransferScheduleWaitList.kt [hunk 2]
```diff
@@ -70,7 +70,7 @@ class OffenderTransferScheduleWaitList(
   @Convert(converter = YesNoConverter::class)
   var approvedFlag: Boolean = false,
 
-  @ManyToOne(fetch = FetchType.LAZY)
+  @ManyToOne(fetch = FetchType.EAGER)
   @JoinColumn(name = "APPROVED_STAFF_ID")
   @NotFound(action = NotFoundAction.IGNORE)
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderTransferScheduleWaitList.kt:75]: Despite an eager and not found, Hibernate won't join on the table hence the fetch join above

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/movements/transfers/offender/OffenderTransferMovementsService.kt [hunk 3]
```diff
@@ -18,14 +20,20 @@ import uk.gov.justice.digital.hmpps.nomisprisonerapi.movements.transfers.schedul
...
+      throw NotFoundException("Offender with nomsId=$offenderNo not found")
+    }
+
     val allSchedules = scheduleRepository.findAllByOffenderBooking_Offender_NomsId(offenderNo)
     val allMovements = movementRepository.findAllByOffenderBooking_Offender_NomsId(offenderNo)
       .filterNot { it.createUsername == "SYS" && it.auditModuleName == "MERGE" }
-    val unscheduledMovements =
-      allMovements.filter { it.transferScheduleOutId == null || it.transferScheduleOutId !in (allSchedules.map { it.eventId }) }
+    val unscheduledMovements = (allSchedules.map { it.eventId }).let { allEventIds ->
+      allMovements.filter { it.transferScheduleOutId == null || it.transferScheduleOutId !in (allEventIds) }
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/movements/transfers/offender/OffenderTransferMovementsService.kt:35]: From previous PR - don't map all schedule IDs for every movement.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/repository/OffenderTransferScheduleOutRepository.kt [hunk 4]
```diff
@@ -3,10 +3,19 @@
...
 import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.OffenderTransferScheduleOut
 
 @Repository
 interface OffenderTransferScheduleOutRepository : JpaRepository<OffenderTransferScheduleOut, Long> {
+  @Query(
+    """
+    SELECT o FROM OffenderTransferScheduleOut o
+        LEFT JOIN FETCH o.waitList
+        LEFT JOIN FETCH o.waitList.approvedStaff
+        WHERE o.offenderBooking.offender.nomsId = :offenderNo
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/repository/OffenderTransferScheduleOutRepository.kt:17]: Prevent N+1 queries

**Review by petergphillips — APPROVED** [review]
LGTM
